### PR TITLE
Fix position of IME in wxSTC, on Windows

### PR DESCRIPTION
**Pull request by wangqr**
Before:
img

After:
img

This patch is partially copied from ScintillaWin. It seems that there is some special case handling in ScintillaWin for Korean, but since I cannot understand or type in Korean, I just left them as is without tesing.

This patch requires imm32.lib. I have added it globally (see the 2nd commit in this PR). Is there a better way so that it is only added to wxSTC?

This PR fix #18759

### CODE REVIEW

**Comment by vadz** [comment]
Thanks, this is clearly very useful because I wouldn't even know where to start with fixing this myself and can't test it (I have the same problem you have with Korean with the C and J in CJK too...), but it can't be applied for several reasons:

1. My first problem, which prevents from even looking at this code in any details, is this code copyright. If you've copied it from elsewhere, you must obtain its original author/copyright holder permission to relicense it under wx licence, otherwise we can't use it at all.
1. You've modified the files `stc.h` and `stc.cpp` which are generated, see the comments at the top of these files. You need to make the changes in the corresponding `.in` files instead.
1. New code in `ScintillaWX.cpp` uses C++11 and wouldn't compile with old MSVC versions we still support. This would need to be either de-C++11-ized or, at the very least, disabled when not using MSVC with the required language features support.
1. Always linking with `imm32.lib` is not great, as it will break the build of all the existing projects using wx. For MSVC we can, and almost certainly should, use `#pragma comment(lib)` to include it from the source file itself instead. For gcc we should define `EXTRALIBS_STC` in `build/bakefiles/config.bkl` similarly to what is already dine for OpenGL libraries there. Alternatively, perhaps we could load the functions dynamically using `wxDynamicLibrary` as it's also already done in some places, although it's not really appealing here, knowing that the functions we want to use are always available since Vista.

**Comment by MaartenBent** [comment]
I think it will be the same license as all the other scintilla code then?

For CMake you can link it directly to `stc` like this in `stc/CMakeLists.txt`:
```c++
if(WIN32)
    wx_lib_link_libraries(stc PRIVATE imm32)
endif()
```

**Comment by wangqr** [comment]
1. The ScintillaWin is part of Scintilla. It is originally in `scintilla/win32`, but that folder is not copied to `wxWidgets/src/stc/scintilla` according to `wxWidgets/src/stc/scintilla/README.txt`. Scintilla's Licence is already included in wxWidgets.

2. Will fix that.

3. I'm aware of this. I only copied a minimal part of the original IME handling code, and removed `std::wstring_view` usage from the original code.

4. Thanks for the suggestion.

**Comment by vadz** [comment]
Ah, sorry, I didn't realize it was part of Scintilla. It should be indeed fine then, thanks.

Looking at the code more closely now, I believe we should do at least part of this in `wxWindow` itself... I don't really know much about IME but it seems clear that any window allowing text entry not using a native control should be doing something like this. And that e.g. wxRTC must be broken in the same way too currently.

You don't have to do any refactoring, of course, it would be already good to fix this "just" for wxSTC, but if you _are_ thinking about putting some parts of these changes into `wxWindow`, please don't hesitate proposing them.

**Comment by wangqr** [comment]
Just tried wxRichTextCtrl and it does have the same issue.

**Comment by wangqr** [comment]
Also, this PR does not implement inline IME, and will always use windowed mode regardless of user settings. The windowed / inline IME are not implemented in ScintillaBase but in ScintillaWin. And I have only ported windowed because it is much simpler.

In case you want to test this locally, or just want to know the difference between the 2 modes: To input CJK, we type on standard keyboard and select the characters from a list. Now if we use Chinese Pinyin IME to input `中`, we type `zhong` and select `中` from the list by pressing `1`.

Assume we have "some text" and the cursor is after "some". With windowed IME it looks like:
img
(Screenshot taken from Notepad++)
Notice the "zhong" is covering " text" part because it is in a floating window.

With inline IME it looks like:
img
(Screenshot taken from Microsoft Word)
And now the "zhong" doesn't cover anything, " text" part is shifted right.

To implement windowed IME we only need to tell the system the cursor position at the beginning of IME composition. To implement inline IME, on the other hand, the text ctrl need to do the composition by itself: temporarily put some text in the ctrl, track those text and update / undo when needed, paint indicator so user can tell the difference between temporary text in IME and actual text, paint cursor, etc. There is existing code in ScintillaWin implementing inline IME, but that is much longer than windowed IME and I have no plan porting them at this moment.

**Comment by wangqr** [comment]
The mingw test on AppVeyor cannot quit on failure. It fails in less than 10 minutes and then just stuck there until 1h timeout. See https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/32779234 and https://ci.appveyor.com/project/wxWidgets/wxwidgets/builds/32782035

**Comment by wangqr** [comment]
The cygwin build timedout due to being slow. Can you restart that build and see if there's any actual issue? Thanks.

File: configure.in [hunk 1]
```diff
@@ -7982,7 +7990,6 @@ EXTRALIBS="$LDFLAGS $LDFLAGS_VERSIONING $LIBS $DMALLOC_LIBS"
 EXTRALIBS_XML="$EXPAT_LINK"
 EXTRALIBS_HTML="$MSPACK_LINK"
 EXTRALIBS_MEDIA="$GST_LIBS"
-EXTRALIBS_STC="-lwxscintilla${WX_LIB_FLAVOUR}-${WX_RELEASE}${HOST_SUFFIX}"
```
Comment (vadz) [inline configure.in:7985]: Sorry, I don't understand: why don't we need this any longer?

File: src/stc/ScintillaWX.cpp [hunk 2]
```diff
@@ -844,6 +845,31 @@ sptr_t ScintillaWX::WndProc(unsigned int iMessage, uptr_t wParam, sptr_t lParam)
...
 
+#ifdef __WXMSW__
+      // ScintillaWin
+      case WM_IME_STARTCOMPOSITION: 	// dbcs
+          // TODO: We haven't ported WM_IME_COMPOSITION yet, so always use windowed IME to make it at least work
+          //if (KoreanIME() || imeInteraction == imeInline) {
+          if (false) {
+              return 0;
+          }
+          else {
```
Comment (vadz) [inline src/stc/ScintillaWX.cpp:856]: This is trivial, but `if (false)` is bound to provoke warnings from at least some compilers, so I'd rather remove this code entirely or at least comment out the entire `if/else` part.

File: src/stc/ScintillaWX.cpp [hunk 3]
```diff
@@ -844,6 +845,31 @@ sptr_t ScintillaWX::WndProc(unsigned int iMessage, uptr_t wParam, sptr_t lParam)
       case SCI_GETDIRECTPOINTER:
             return reinterpret_cast<sptr_t>(this);
 
+#ifdef __WXMSW__
+      // ScintillaWin
+      case WM_IME_STARTCOMPOSITION: 	// dbcs
```
Comment (vadz) [inline src/stc/ScintillaWX.cpp:850]:
```suggestion
      case WM_IME_STARTCOMPOSITION:
```
Trivial, but these "dbcs" comments just don't seem helpful to me.

File: src/stc/ScintillaWX.cpp [hunk 4]
```diff
@@ -844,6 +845,31 @@ sptr_t ScintillaWX::WndProc(unsigned int iMessage, uptr_t wParam, sptr_t lParam)
...
+
+      case WM_IME_ENDCOMPOSITION: 	// dbcs
+          ImeEndComposition();
+          return stc->wxControl::MSWWindowProc(iMessage, wParam, lParam);
+
+      case WM_IME_KEYDOWN:
+      case WM_IME_REQUEST:
+      case WM_IME_COMPOSITION:
+      case WM_IME_SETCONTEXT:
+          // TODO: Port event handling from ScintillaWin
```
Comment (vadz) [inline src/stc/ScintillaWX.cpp:869]: I'd like to understand what exactly is meant here. Is it for inline IME support? If so, we should probably say so directly and maybe remove "TODO" as we have no plans for doing it at the moment (and just say that we'd need to implement this for inline IME).

**Review by vadz — COMMENTED** [review]
Thanks for the changes and also for the explanation of the difference between window/inline IME modes, this was helpful!

I definitely agree that we can live without the inline mode, it is "just" a convenience, unlike the basic IME support which is an absolute necessity.

So I think we should merge this already and will do it soon if there are no objections. The only thing I'm not sure about is the change in configure commented on below.

File: configure.in (hunk 1 above)
Comment (wangqr) [reply configure.in:7985]: This variable is only set but never used before. You can search for `EXTRALIBS_STC` in Makefile.in and there is no result. It is used in order to link against imm32 after the change I made in:
https://github.com/wxWidgets/wxWidgets/pull/1852/commits/f730da3c14927b1833fc776479bd457b494bb30e#diff-0c9f5c5268aea486bb1cf4ce87975c38R414

STC does require wxScintilla, but it is already done in https://github.com/wxWidgets/wxWidgets/blob/master/build/bakefiles/multilib.bkl#L395

File: src/stc/ScintillaWX.cpp (hunk 4 above)
Comment (wangqr) [reply src/stc/ScintillaWX.cpp:869]: Yes, it is for inline IME

File: src/stc/ScintillaWX.cpp (hunk 3 above)
Comment (wangqr) [reply src/stc/ScintillaWX.cpp:850]: This, and the `if (false)` above, are just trying to make minimal changes to original SciWin code. Should I remove them for better readability or comment it out for minimal change?

**Comment by vadz** [comment]
Thanks again for fixing this and sorry for the delay with merging it!

**Pull request merged**